src/core/xds/grpc/xds_matcher_parse.cc

```diff
@@ -0,0 +1,354 @@
+}
+
+// Parse OnMatch components of the matcher
+// TODO(bpawan): b/431645556 this should have a max recusrsion depth
+std::optional<XdsMatcher::OnMatch> ParseOnMatch(
+    const XdsResourceType::DecodeContext& context,
+    const xds_type_matcher_v3_Matcher_OnMatch* on_match,
+    const XdsMatcherActionRegistry& action_registry,
+    const UniqueTypeName& matcher_context, ValidationErrors* errors) {
+  // TODO(bpawan): b/431645620 Parse keep matching once we move to latest xds
+  // protos
+  bool keep_matching = false;
+  // Action is a variant which can have Action or a Nested Matcher
+  if (xds_type_matcher_v3_Matcher_OnMatch_has_action(on_match)) {
+    ValidationErrors::ScopedField field(errors, ".action");
+    auto action = ParseAction(
+        context, xds_type_matcher_v3_Matcher_OnMatch_action(on_match),
+        action_registry, errors);
+    return XdsMatcher::OnMatch(std::move(action), keep_matching);
+  } else if (xds_type_matcher_v3_Matcher_OnMatch_has_matcher(on_match)) {
+    ValidationErrors::ScopedField field(errors, ".matcher");
+    auto nested_matcher = ParseXdsMatcher(
+        context, xds_type_matcher_v3_Matcher_OnMatch_matcher(on_match),
+        action_registry, matcher_context, errors);
+    return XdsMatcher::OnMatch(std::move(nested_matcher), keep_matching);
+  } else {
+    errors->AddError("Unknown field in OnMatch");
+  }
+  return std::nullopt;
+}
+
+// Parse MatchTree Map
+absl::flat_hash_map<std::string, XdsMatcher::OnMatch> ParseMatchMap(
+    const XdsResourceType::DecodeContext& context,
+    const xds_type_matcher_v3_Matcher_MatcherTree_MatchMap* match_map,
+    const XdsMatcherActionRegistry& action_registry,
+    const UniqueTypeName& matcher_context, ValidationErrors* errors) {
+  absl::flat_hash_map<std::string, XdsMatcher::OnMatch> result;
+  if (xds_type_matcher_v3_Matcher_MatcherTree_MatchMap_map_size(match_map) ==
+      0) {
+    errors->AddError("map is empty");
+    return result;
+  }
+  auto iter = kUpb_Map_Begin;
+  upb_StringView upb_key;
+  const xds_type_matcher_v3_Matcher_OnMatch* value;
+  while (xds_type_matcher_v3_Matcher_MatcherTree_MatchMap_map_next(
+      match_map, &upb_key, &value, &iter)) {
+    ValidationErrors::ScopedField field(errors, ".on_match");
+    auto on_match =
+        ParseOnMatch(context, value, action_registry, matcher_context, errors);
+    if (on_match.has_value()) {
+      result.emplace(UpbStringToStdString(upb_key), std::move(*on_match));
+    }
+  }
+  return result;
+}
+
+// Parse SinglePredicate
+std::unique_ptr<XdsMatcherList::Predicate> ParseSinglePredicate(
+    const XdsResourceType::DecodeContext& context,
+    const xds_type_matcher_v3_Matcher_MatcherList_Predicate_SinglePredicate*
+        single_predicate,
+    const UniqueTypeName& matcher_context, ValidationErrors* errors) {
+  // Supporting value match now, need to add custom match
+  if (!xds_type_matcher_v3_Matcher_MatcherList_Predicate_SinglePredicate_has_value_match(
```

Comment:
Suggest changing this to:

```
std::unique_ptr<XdsMatcherList::InputMatcher<absl::string_view>>
    input_string_matcher;
{
  ValidationErrors::ScopedField field(errors, ".value_match");
  const auto* value_match =
      xds_type_matcher_v3_Matcher_MatcherList_Predicate_SinglePredicate_value_match
          single_predicate);
  if (value_match == nullptr) {
    errors->AddError("field not present");
    return nullptr;
  }
  input_string_matcher = ParseStringMatcher(context, value_match, errors);
}
```

This avoids accessing the `value_match` field twice.  It also provides a better error message when the field is not present.